File: ruby/lib/fizzy/auth_strategy.rb
# frozen_string_literal: true

module Fizzy
  # AuthStrategy controls how authentication is applied to HTTP requests.
  # The default strategy is BearerAuth, which uses a TokenProvider to set
  # the Authorization header with a Bearer token.
  #
  # Custom strategies can implement alternative auth schemes such as
  # cookie-based auth or magic link flows.
  #
  # To implement a custom strategy, create a class that responds to
  # #authenticate(headers), where headers is a Hash that you can modify.
  module AuthStrategy
    # Apply authentication to the given headers hash.
    # @param headers [Hash] the request headers to modify
    def authenticate(headers)
      raise NotImplementedError, "#{self.class} must implement #authenticate"
    end
  end

  # Bearer token authentication strategy (default).
  # Sets the Authorization header with "Bearer {token}".
  class BearerAuth
    include AuthStrategy

    # @param token_provider [TokenProvider] provides access tokens
    def initialize(token_provider)
      @token_provider = token_provider
    end

    # @return [TokenProvider] the underlying token provider
    attr_reader :token_provider

    def authenticate(headers)
      headers["Authorization"] = "Bearer #{@token_provider.access_token}"
    end
  end
end

File: typescript/src/auth-strategy.ts
/**
 * Authentication strategy for the Fizzy SDK.
 *
 * AuthStrategy controls how authentication is applied to HTTP requests.
 * Fizzy supports Bearer tokens and Cookie sessions.
 */

import type { TokenProvider } from "./client.js";

/**
 * AuthStrategy controls how authentication is applied to HTTP requests.
 * Called before every HTTP request to apply credentials to headers.
 */
export interface AuthStrategy {
  /**
   * Apply authentication to the given request headers.
   * Called before every HTTP request.
   */
  authenticate(headers: Headers): Promise<void>;
}

/**
 * Bearer token authentication strategy.
 * Sets the Authorization header with "Bearer {token}".
 */
export class BearerAuth implements AuthStrategy {
  private tokenProvider: TokenProvider;

  constructor(tokenProvider: TokenProvider) {
    this.tokenProvider = tokenProvider;
  }

  async authenticate(headers: Headers): Promise<void> {
    const token =
      typeof this.tokenProvider === "function"
        ? await this.tokenProvider()
        : this.tokenProvider;
    headers.set("Authorization", `Bearer ${token}`);
  }
}

/**
 * Creates a BearerAuth strategy from a TokenProvider.
 */
export function bearerAuth(tokenProvider: TokenProvider): AuthStrategy {
  return new BearerAuth(tokenProvider);
}
